python/ray/data/_internal/util.py

```diff
@@ -206,3 +208,51 @@ def _check_import(obj, *, module: str, package: str) -> None:
             f"couldn't be imported. You can install '{package}' by running `pip "
             f"install {package}`."
         )
+
+
+def _resolve_custom_scheme(path: str) -> str:
+    """Returns the resolved path if the given path following a Ray-specific custom
+    scheme. Othewise, returns the path unchanged.
+
+    The supported custom schemes are: "local", "example".
+    """
+    import pathlib
+    import urllib.parse
+
+    parsed_uri = urllib.parse.urlparse(path)
+    if parsed_uri.scheme == _LOCAL_SCHEME:
+        path = parsed_uri.netloc + parsed_uri.path
+    elif parsed_uri.scheme == _EXAMPLE_SCHEME:
+        example_data_path = pathlib.Path(__file__).parent.parent / "examples" / "data"
+        path = example_data_path / parsed_uri.netloc
+        path = str(path.resolve())
+    return path
+
+
+def _is_local_scheme(paths: Union[str, List[str]]) -> bool:
+    """Returns True if the given paths are in local scheme.
+    Note: The paths must be in same scheme, i.e. it's invalid and
+    will raise error if paths are mixed with different schemes.
+    """
+    import pathlib
+    import urllib.parse
+
+    if isinstance(paths, str):
+        paths = [paths]
+    if isinstance(paths, pathlib.Path):
+        paths = [str(paths)]
+    elif not isinstance(paths, list) or any(not isinstance(p, str) for p in paths):
+        raise ValueError("paths must be a path string or a list of path strings.")
+    elif len(paths) == 0:
+        raise ValueError("Must provide at least one path.")
+    is_local = None
+    for path in paths:
+        current = urllib.parse.urlparse(path).scheme == _LOCAL_SCHEME
+        if not is_local:
```

Comment:
This will always reset `is_local is False` to whatever `current` is, which I don't think is what we want, right? E.g. this seems like it would return `True` for `["s3://foo/bar", "local://foo/bar"]` rather than raising an error.
```suggestion
        if is_local is not None:
```

Reply:
Good catch!